src/System.Private.Xml.Linq/src/System/Xml/XPath/XNodeNavigator.cs

```diff
@@ -878,9 +878,9 @@ private IEnumerable<T> EvaluateIterator<T>(XPathNodeIterator result)
                 XText t = r as XText;
                 if (t != null && t.GetParent() != null)
                 {
-                    foreach (XNode node in t.GetParent().Nodes())
+                    while (t != t.parent.content)
                     {
-                        t = node as XText;
+                        t = t.next as XText;
```

Comment:
if the only diff was that this was returning also itself - wouldn't adding `if (node != this)` be equivalent then? the Desktop code is depending a lot on the internals of other classes, the internals make it harder to read this code, I prefer when it uses public APIs but also from factoring perspective (on .NET Core I changed this code because we split XPath from XDocument and the internals were not accessible anymore and using internals is generally a bad practice)